File: Orimath/ViewModels/PluginSettingViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Reflection;
using System.Windows.Input;
using Orimath.Controls;
using Orimath.Plugins;
using ApplicativeProperty;
using Sssl;

namespace Orimath.ViewModels
{
    public class PluginSettingViewModel : NotifyPropertyChanged
    {
        private readonly PluginLoadSettingViewModel _loadSetting;

        public PluginSettingPageViewModel[] Pages { get; }

        public PluginSettingViewModel(IMessenger messenger)
        {
            _loadSetting = new PluginLoadSettingViewModel(messenger);

            Pages =
                Enumerable.Concat(
                    new PluginSettingPageViewModel[] { _loadSetting },
                    PluginExecutor.ConfigurablePlugins.Select(c => new PluginItemSettingViewModel(c)))
                .ToArray();

            SaveCommand = Prop.True.ToCommand(_loadSetting.Save);
            CloseCommand = messenger.CloseDialogCommand;
        }

        public ICommand SaveCommand { get; }

        public ICommand CloseCommand { get; }
     }

    public abstract class PluginSettingPageViewModel : NotifyPropertyChanged
    {
        public abstract string Header { get; }
    }

    public class PluginLoadSettingViewModel : PluginSettingPageViewModel
    {
        private readonly IMessenger _messenger;

        public override string Header => "ON/OFFと読み込み順";

        public ObservableCollection<PluginViewModel> Plugins { get; } = new();

        public PluginLoadSettingViewModel(IMessenger messenger)
        {
            _messenger = messenger;

            SetViewModels(PluginExecutor.LoadedPluginTypes.Concat(PluginExecutor.LoadedViewPluginTypes).ToArray(),
                          PluginExecutor.Setting.PluginOrder, Plugins);

            PluginIndex = new ValueProp<int>(0);
            UpPluginCommand = PluginIndex.Select(i => 1 <= i && i < Plugins.Count).ToCommand(UpPlugin);
            DownPluginCommand = PluginIndex.Select(i => 0 <= i && i < Plugins.Count - 1).ToCommand(DownPlugin);
        }

        private static void SetViewModels(Type[] pluginTypes, string[] order, ObservableCollection<PluginViewModel> collection)
        {
            var plugins = pluginTypes
                .Select(t => new PluginViewModel(t, false))
                .ToDictionary(vm => vm.FullName);

            foreach (var fullName in order)
            {
                if (plugins.TryGetValue(fullName, out var plugin))
                {
                    plugin.IsEnabled = true;
                    plugins.Remove(fullName);
                    collection.Add(plugin);
                }
            }

            foreach(var plugin in  plugins.Values)
            {
                collection.Add(plugin);
            }
        }

        public ValueProp<int> PluginIndex { get; }

        public void Save(object? dummy)
        {
            PluginExecutor.Setting.PluginOrder = Plugins.Where(x => x.IsEnabled).Select(x => x.FullName).ToArray();
            foreach (var configurable in PluginExecutor.ConfigurablePlugins)
                PluginExecutor.Setting.Settings[configurable.GetType().ToString()] = SsslObject.ConvertFrom(configurable.Setting);
            PluginExecutor.SaveSetting();
            _messenger.CloseDialog();
        }

        public void UpPlugin(object? dummy) => Plugins.Move(PluginIndex.Value, PluginIndex.Value - 1);

        public void DownPlugin(object? dummy) => Plugins.Move(PluginIndex.Value, PluginIndex.Value + 1);

        public ICommand UpPluginCommand { get; }

        public ICommand DownPluginCommand { get; }
    }

    public class PluginItemSettingViewModel : PluginSettingPageViewModel
    {
        public override string Header { get; }

        public SettingViewModel Content { get; }

        public PluginItemSettingViewModel(IConfigurablePlugin plugin)
        {
            var pluginType = plugin.GetType();
            Header = pluginType.GetCustomAttribute<DisplayNameAttribute>()?.DisplayName ?? pluginType.Name;
            
            Content = new SettingViewModel(plugin.Setting);
        }
    }

    public class PluginViewModel : NotifyPropertyChanged
    {
        public string Name { get; }
        public string Description { get; }
        public string FullName { get; }
        public Type Type { get; }

        private bool _isEnabled;
        public bool IsEnabled { get => _isEnabled; set => SetValue(ref _isEnabled, value); }

        public PluginViewModel(Type pluginType, bool isEnabled)
        {
            Name = pluginType.GetCustomAttribute<DisplayNameAttribute>()?.DisplayName ?? pluginType.Name;
            Description = pluginType.GetCustomAttribute<DescriptionAttribute>()?.Description ?? "(No description)";

            FullName = pluginType.FullName!;
            Type = pluginType;
            _isEnabled = isEnabled;
        }
    }
}

File: Orimath/ViewModels/PluginSettingViewModel.fs
namespace Orimath.ViewModels
open System
open System.Collections.ObjectModel
open System.ComponentModel
open System.Linq
open System.Reflection
open Orimath
open Orimath.Controls
open Orimath.Plugins
open Orimath.Internal
open ApplicativeProperty
open Sssl

[<AbstractClass>]
type PluginSettingPageViewModel() =
    inherit NotifyPropertyChanged()

    abstract member Header : string


type PluginViewModel(pluginType: Type, isEnabled: bool) =
    inherit NotifyPropertyChanged()

    member val IsEnabled = Prop.value isEnabled
    member _.FullName = pluginType.FullName
    member _.Type = pluginType

    member val Name =
        pluginType.GetCustomAttribute<DisplayNameAttribute>()
        |> Null.bind(fun att -> att.DisplayName)
        |> Null.defaultValue pluginType.Name

    member val Description =
        pluginType.GetCustomAttribute<DescriptionAttribute>()
        |> Null.bind(fun att -> att.Description)
        |> Null.defaultValue "(No description)"



type PluginItemSettingViewModel(plugin: IConfigurablePlugin, dispatcher: IDispatcher) =
    inherit PluginSettingPageViewModel()
    let pluginType = plugin.GetType()
    let header =
        pluginType.GetCustomAttribute<DisplayNameAttribute>()
        |> Null.bind(fun att -> att.DisplayName)
        |> Null.defaultValue pluginType.Name

    override _.Header = header
    member val Content = SettingViewModel(plugin.Setting, dispatcher)


type PluginLoadSettingViewModel(messenger: IMessenger) =
    inherit PluginSettingPageViewModel()
    let plugins = ObservableCollection<PluginViewModel>()
    let pluginIndex = Prop.value 0
    do
        let pluginMap =
            (Seq.append PluginExecutor.loadedPluginTypes PluginExecutor.loadedViewPluginTypes)
             .ToDictionary((fun t -> t.FullName), (fun t -> PluginViewModel(t, false)))
        let order = PluginExecutor.setting.PluginOrder
        for fullName in order do
            match pluginMap.TryGetValue(fullName) with
            | BoolSome(plugin) ->
                plugin.IsEnabled.Value <- true
                ignore (pluginMap.Remove(fullName))
                plugins.Add(plugin)
            | BoolNone -> ()
        for plugin in pluginMap.Values do
            plugins.Add(plugin)

    member _.Plugins = plugins
    member _.PluginIndex = pluginIndex
    
    member val UpPluginCommand =
        pluginIndex
        |> Prop.map(fun i -> 1 <= i && i < plugins.Count)
        |> Prop.command(fun _ -> plugins.Move(pluginIndex.Value, pluginIndex.Value - 1))
    
    member val DownPluginCommand =
        pluginIndex
        |> Prop.map(fun i -> 0 <= i && i < plugins.Count - 1)
        |> Prop.command(fun _ -> plugins.Move(pluginIndex.Value, pluginIndex.Value + 1))

    override _.Header = "ON/OFFと読み込み順"

    member _.Save(_: obj) =
        PluginExecutor.setting.PluginOrder <-
            [| for x in plugins do if x.IsEnabled.Value then yield x.FullName |]
        for configurable in PluginExecutor.configurablePlugins do
            PluginExecutor.setting.Settings.[configurable.GetType().ToString()] <-
                SsslObject.ConvertFrom(configurable.Setting)
        PluginExecutor.saveSetting()
        messenger.CloseDialog()


type PluginSettingViewModel(messenger: IMessenger, dispatcher: IDispatcher) =
    inherit NotifyPropertyChanged()
    let loadSetting = PluginLoadSettingViewModel(messenger)

    member val Pages =
        [| yield loadSetting :> PluginSettingPageViewModel
           for c in PluginExecutor.configurablePlugins do
               yield upcast PluginItemSettingViewModel(c, dispatcher) |]

    member val SaveCommand = Prop.ctrue |> Prop.command loadSetting.Save

    member _.CloseCommand = messenger.CloseDialogCommand